source/slang/slang-ir.cpp

```diff
@@ -857,9 +857,23 @@ void fixUpFuncType(IRFunc* func, IRType* resultType)
     builder.setInsertBefore(func);
 
     List<IRType*> paramTypes;
-    for (auto param : func->getParams())
+    // As for extern function, it's not definition, that is no params. After legalize SPIR-V, the
+    // function type hasn't params, it's a error signature. When we fix the function type, extern
+    // function should check the param types in function type instead of params.
+    if (func->isDefinition())
     {
-        paramTypes.add(param->getFullType());
+        for (auto param : func->getParams())
+        {
+            paramTypes.add(param->getFullType());
+        }
+    }
+    else
+    {
+        auto funcType = as<IRFuncType>(func->getFullType());
```

Comment:
There are two issues here:

* This code almost certainly should be using `func->getDataType()` instead of `func->getFullType()`. The "full type" of an instruction can include wrapper types that applies rates, attributes, etc. and this logic basically wants to look through any/all of that.

* There is nothing being done here to check the result of the `as` operation. This logic could either assert that `funcType` is non-null, or else use `cast<IRFuncType>(...)` instead of `as<IRFuncType>(...)`, since the `cast` function includes its own assertion. While it is reasonable to expect a function to have a function type, that doesn't mean we shouldn't be at least a little defensive against the possibility that the IR is in an invalid state.